File: src/analysis/E/engine_heterogeneity.py
import argparse
import sys
import pandas as pd
import numpy as np
import os
import json
import re
from pathlib import Path
import datetime
import statsmodels.formula.api as smf
from statsmodels.stats.multitest import multipletests
from linearmodels.panel import PanelOLS

# Add project root to path
project_root = str(Path(__file__).resolve().parents[3])
if project_root not in sys.path:
    sys.path.append(project_root)

# Import Helpers (Reusing D helpers as they are compatible stats tools)
# Import Helpers (Reusing D helpers)
from src.analysis.D.helpers.stat_utils import validate_and_prep_data, run_r_wrapper
from src.helpers.data_loader import load_analysis_metrics

# --- CONFIGURATION (Locked per Plan v10) ---

metrics = load_analysis_metrics()
BLOCK_SEMANTICS = metrics.get("semantic", [])
BLOCK_READABILITY = metrics.get("readability", [])
BLOCK_PERFORMANCE = metrics.get("performance", [])
BLOCK_ACCESSIBILITY = metrics.get("accessibility", [])

# Primary predictors
HETEROGENEITY_PREDICTORS = BLOCK_SEMANTICS + BLOCK_PERFORMANCE + BLOCK_ACCESSIBILITY + BLOCK_READABILITY

def generate_replication_grid(coeffs, predictors):
    """
    Analyzes engine-stratified coefficients to generate a Replication Grid.
    Metrics:
    - Sign Agreement (All signs match?)
    - CI Overlap (Do 95% CIs overlap across all engines?)
    - Practical Significance Stability (Is practical_flag consistent?)
    """
    grid_stats = []

    # Determine the complete set of engines represented in this subset. A term
    # can be called cross-engine consistent only when it was estimated for every
    # engine; agreement across an incomplete set must not count as replication.
    expected_engines = set()
    for row in coeffs:
        match = re.match(r"^RQ6_Stratified_(.*?)_R_", str(row.get("model_id", "")))
        if match:
            expected_engines.add(match.group(1))
    
    for term in predictors:
        # Get coeffs for this term from engine models
        relevant = [c for c in coeffs if c['term'] == term and str(c.get('model_id', '')).startswith('RQ6_Stratified_')]
        
        if not relevant: continue

        term_engines = set()
        for row in relevant:
            match = re.match(r"^RQ6_Stratified_(.*?)_R_", str(row.get("model_id", "")))
            if match:
                term_engines.add(match.group(1))
        complete_engine_coverage = (
            bool(expected_engines)
            and term_engines == expected_engines
            and len(relevant) == len(expected_engines)
        )
        
        signs = [np.sign(r.get('effect_size', r.get('coef'))) for r in relevant]
        # Use p_fdr if available, fallback to p_raw/pval
        pvals = [r.get('p_fdr', r.get('p_raw', r.get('pval', 1.0))) for r in relevant]
        cis = [(r.get('ci_lower_95', r.get('ci_lower')), r.get('ci_upper_95', r.get('ci_upper'))) for r in relevant]
        
        # Agreement: All signs same?
        all_same_sign = complete_engine_coverage and len(set(signs)) == 1
        
        # Significant Agreement: All FDR-significant AND same sign?
        sig_agreement = complete_engine_coverage and all(pd.notna(p) and p < 0.05 for p in pvals) and all_same_sign
        
        # CI Overlap
        max_lower = max([c[0] for c in cis])
        min_upper = min([c[1] for c in cis])
        ci_overlap = complete_engine_coverage and max_lower <= min_upper
        
        stats = {
            'term': term,
            'n_engines': len(relevant),
            'all_sign_agreement': all_same_sign,
            'significant_agreement': sig_agreement,
            'ci_overlap': ci_overlap,
            'median_coef': np.median([r.get('effect_size', r.get('coef')) for r in relevant]),
            'mad_coef': np.median(np.abs(np.array([r.get('effect_size', r.get('coef')) for r in relevant]) - np.median([r.get('effect_size', r.get('coef')) for r in relevant])))
        }
        grid_stats.append(stats)
        
    return grid_stats

def main():
    parser = argparse.ArgumentParser(description="Analysis E (RQ6): Engine Heterogeneity (R Wrapper)")
    parser.add_argument("--dataset", default="data/dataset.parquet", help="Path to merged dataset")
    parser.add_argument("--out-dir", default="data/analysis/E", help="Output directory")
    parser.add_argument("--dataset-variant", default="clean", help="Variant tag for dataset")
    parser.add_argument("--code-version", default="unknown", help="Git SHA or version tag")
    args = parser.parse_args()
    
    os.makedirs(args.out_dir, exist_ok=True)
    
    # Validation only (R script assumes clean data mostly, but we validate path)
    if not os.path.exists(args.dataset):
        sys.exit(f"Dataset not found: {args.dataset}")
        
    print("Running RQ6 Engine Heterogeneity Analysis (via R)...", file=sys.stderr)
    
    r_script = os.path.join(project_root, "src/analysis/E/engine_heterogeneity.R")
    success = run_r_wrapper(r_script, ["--dataset", args.dataset, "--out-dir", args.out_dir])
    
    if not success:
        sys.exit("R script execution failed.")
        
    # --- Post-Processing: Generate Replication Grid ---
    r_out_csv = os.path.join(args.out_dir, "heterogeneity_coeffs_r.csv")
    if os.path.exists(r_out_csv):
        print("Processing R output for Replication Grid...", file=sys.stderr)
        try:
            df_coeffs = pd.read_csv(r_out_csv)
            coeffs_list = df_coeffs.to_dict('records')
            
            # Find unique subsets for stratified models
            stratified_coeffs = [c for c in coeffs_list if str(c.get('model_id', '')).startswith('RQ6_Stratified_')]
            # Subset values look like: Engine_google_Full, Engine_google_NoSource
            # Let's extract the actual data subset (Full, NoSource, Source)
            # The model_id usually ends with _Full, _NoSource, or _Source
            subsets = set()
            for c in stratified_coeffs:
                m_id = str(c.get('model_id', ''))
                match = re.search(r'_(Full|NoSource|Source)$', m_id)
                if match:
                    subsets.add(match.group(1))
            
            if not subsets:
                subsets = {'Full'} # Fallback
                
            all_grid_stats = []
            for subset in subsets:
                # Filter coeffs for this specific subset
                subset_coeffs = [c for c in stratified_coeffs if str(c.get('model_id', '')).endswith(f"_{subset}")]
                if not subset_coeffs: continue
                
                grid_stats = generate_replication_grid(subset_coeffs, HETEROGENEITY_PREDICTORS)
                for g in grid_stats:
                    g['subset'] = subset
                all_grid_stats.extend(grid_stats)
            
            if all_grid_stats:
                pd.DataFrame(all_grid_stats).to_csv(f"{args.out_dir}/replication_grid.csv", index=False)
                print("Generated replication_grid.csv", file=sys.stderr)
            
        except Exception as e:
            print(f"Error processing replication grid: {e}", file=sys.stderr)
    else:
        print("Warning: R output CSV not found. Skipping replication grid.", file=sys.stderr)

    print(f"\nSaved results to {args.out_dir}", file=sys.stderr)

if __name__ == "__main__":
    main()

File: src/analysis/E/engine_heterogeneity.R
#!/usr/bin/env Rscript

# Dependencies: install.packages(c("fixest", "data.table", "argparse", "broom"))
# This script implements Analysis E (RQ6): Engine Heterogeneity

suppressPackageStartupMessages({
    library(fixest)
    library(data.table)
    library(argparse)
    library(arrow)
})

# Find and source data_loader.R
source_helper <- function(relative_path) {
    paths <- c(relative_path, file.path("..", relative_path), file.path("../..", relative_path), file.path("../../..", relative_path))
    for (p in paths) {
        if (file.exists(p)) {
            source(p)
            return(invisible())
        }
    }
    stop(paste("Could not find helper:", relative_path))
}

invisible(source_helper("src/helpers/data_loader.R"))

parser <- ArgumentParser(description = "Analysis E: Engine Heterogeneity (RQ6) (R Implementation)")
parser$add_argument("--dataset", type = "character", required = TRUE, help = "Path to prepared Parquet dataset")
parser$add_argument("--out-dir", type = "character", required = TRUE, help = "Output directory")
args <- parser$parse_args()

# Load Data
if (!file.exists(args$dataset)) {
    stop(paste("Dataset not found:", args$dataset))
}

cat("Loading dataset (Parquet):", args$dataset, "\n")
df <- load_analysis_dataset(args$dataset)

# Ensure output directory
if (!dir.exists(args$out_dir)) {
    dir.create(args$out_dir, recursive = TRUE)
}

# Ensure types
# Start by checking if 'recip_rank' exists. If not, maybe it needs standardizing?
if (!"recip_rank" %in% names(df) && "rank" %in% names(df)) {
    df[, recip_rank := 1 / rank]
}

if (!"recip_rank" %in% names(df)) {
    stop("Column 'recip_rank' not found in dataset")
}

df[, `:=`(
    search_engine = as.factor(as.character(search_engine)),
    search_term = as.factor(as.character(search_term)),
    recip_rank = as.numeric(recip_rank)
)]

# Define Variables
metrics <- load_analysis_metrics()
semantics <- metrics$semantic
readability <- metrics$readability
performance <- metrics$performance
accessibility <- metrics$accessibility

core_predictors <- c(semantics, readability, performance, accessibility)

# Standardize Continuous Predictors (for practical_flag threshold comparability)
cols_to_scale <- core_predictors[core_predictors %in% names(df)]
if (length(cols_to_scale) > 0) {
    for (v in cols_to_scale) {
        df[, (v) := as.numeric(scale(df[[v]]))]
    }
    cat("Standardized", length(cols_to_scale), "continuous predictors.\n")
}

# --- RQ6: Pooled Interaction Model ---
# Model: recip_rank ~ (Core Predictors) * search_engine + FE
# fixest syntax for interaction: x1:search_engine or i(search_engine, x1)

cat("Running Pooled Interaction Model...\n")
interaction_terms <- paste0("i(search_engine, ", core_predictors, ", ref='google')")
formula_str <- paste(
    "recip_rank ~",
    paste(core_predictors, collapse = " + "),
    "+",
    paste(interaction_terms, collapse = " + "),
    "+ i(search_engine, ref='google')", # Main effect of engine
    "| search_term"
)
f_interaction <- as.formula(formula_str)

results_list <- list()

tryCatch(
    {
        # Full model
        mod_int_full <- feols(f_interaction, data = df, cluster = ~search_term)

        # Extract
        res_table <- coeftable(mod_int_full)
        res_dt <- as.data.table(res_table, keep.rownames = "term")
        setnames(res_dt, c("term", "coef", "se", "tstat", "pval"))

        res_dt[, `:=`(
            model_id = "RQ6_Interaction_R_Full",
            n_obs = nobs(mod_int_full),
            r2 = r2(mod_int_full, "r2"),
            r2_within = r2(mod_int_full, "wr2"),
            software = "R_fixest",
            subset = "Full"
        )]

        ci <- confint(mod_int_full)
        res_dt[, `:=`(ci_lower = ci[, 1], ci_upper = ci[, 2])]

        results_list[["Interaction_Full"]] <- res_dt

        # Split models
        if ("is_source_domain" %in% names(df)) {
            mod_int_split <- feols(f_interaction, data = df, cluster = ~search_term, split = ~is_source_domain)
            for (i in seq_along(mod_int_split)) {
                mod <- mod_int_split[[i]]
                res_tab <- coeftable(mod)
                res_dt_split <- as.data.table(res_tab, keep.rownames = "term")
                setnames(res_dt_split, c("term", "coef", "se", "tstat", "pval"))

                split_val <- names(mod_int_split)[i]
                subset_name <- ifelse(grepl("FALSE", split_val, ignore.case = TRUE), "NoSource",
                    ifelse(grepl("TRUE", split_val, ignore.case = TRUE), "Source", split_val)
                )

                res_dt_split[, `:=`(
                    model_id = paste0("RQ6_Interaction_R_", subset_name),
                    n_obs = nobs(mod),
                    r2 = r2(mod, "r2"),
                    r2_within = r2(mod, "wr2"),
                    software = "R_fixest",
                    subset = subset_name
                )]

                ci_split <- confint(mod)
                res_dt_split[, `:=`(ci_lower = ci_split[, 1], ci_upper = ci_split[, 2])]

                results_list[[paste0("Interaction_", subset_name)]] <- res_dt_split
            }
        }
    },
    error = function(e) {
        cat("Error in Interaction Model:", e$message, "\n")
    }
)

# --- RQ6: Engine-Stratified Models ---
engines <- unique(as.character(df$search_engine))

for (eng in engines) {
    cat("Running Stratified Model for:", eng, "...\n")
    df_sub <- df[search_engine == eng]

    f_strat <- as.formula(paste(
        "recip_rank ~", paste(core_predictors, collapse = " + "), "| search_term"
    ))

    tryCatch(
        {
            mod_strat_full <- feols(f_strat, data = df_sub, cluster = ~search_term)

            res_table <- coeftable(mod_strat_full)
            res_dt <- as.data.table(res_table, keep.rownames = "term")
            setnames(res_dt, c("term", "coef", "se", "tstat", "pval"))

            res_dt[, `:=`(
                model_id = paste0("RQ6_Stratified_", eng, "_R_Full"),
                n_obs = nobs(mod_strat_full),
                r2 = r2(mod_strat_full, "r2"),
                r2_within = r2(mod_strat_full, "wr2"),
                software = "R_fixest",
                subset = paste0("Engine_", eng, "_Full")
            )]

            ci <- confint(mod_strat_full)
            res_dt[, `:=`(ci_lower = ci[, 1], ci_upper = ci[, 2])]

            results_list[[paste0("Stratified_", eng, "_Full")]] <- res_dt

            # Split stratified models
            if ("is_source_domain" %in% names(df_sub) && length(unique(df_sub$is_source_domain)) > 1) {
                mod_strat_split <- feols(f_strat, data = df_sub, cluster = ~search_term, split = ~is_source_domain)
                for (i in seq_along(mod_strat_split)) {
                    mod <- mod_strat_split[[i]]
                    res_tab <- coeftable(mod)
                    res_dt_split <- as.data.table(res_tab, keep.rownames = "term")
                    setnames(res_dt_split, c("term", "coef", "se", "tstat", "pval"))

                    split_val <- names(mod_strat_split)[i]
                    split_name <- ifelse(grepl("FALSE", split_val, ignore.case = TRUE), "NoSource",
                        ifelse(grepl("TRUE", split_val, ignore.case = TRUE), "Source", split_val)
                    )

                    res_dt_split[, `:=`(
                        model_id = paste0("RQ6_Stratified_", eng, "_R_", split_name),
                        n_obs = nobs(mod),
                        r2 = r2(mod, "r2"),
                        r2_within = r2(mod, "wr2"),
                        software = "R_fixest",
                        subset = paste0("Engine_", eng, "_", split_name)
                    )]

                    ci_split <- confint(mod)
                    res_dt_split[, `:=`(ci_lower = ci_split[, 1], ci_upper = ci_split[, 2])]
                    results_list[[paste0("Stratified_", eng, "_", split_name)]] <- res_dt_split
                }
            } else if ("is_source_domain" %in% names(df_sub)) {
                # Only one value exists (either only source or only nosource found)
                split_val <- df_sub$is_source_domain[1]
                split_name <- ifelse(split_val == FALSE, "NoSource", "Source")
                res_dt_copy <- copy(res_dt)
                res_dt_copy[, `:=`(
                    model_id = paste0("RQ6_Stratified_", eng, "_R_", split_name),
                    subset = paste0("Engine_", eng, "_", split_name)
                )]
                results_list[[paste0("Stratified_", eng, "_", split_name)]] <- res_dt_copy
            }
        },
        error = function(e) {
            cat("Error in Stratified Model", eng, ":", e$message, "\n")
        }
    )
}

# Bind and Save
# Bind and Save
if (length(results_list) > 0) {
    all_results <- rbindlist(results_list, fill = TRUE)

    # Standardize (Plan 4.1)
    setnames(all_results,
        old = c("coef", "ci_lower", "ci_upper", "pval"),
        new = c("effect_size", "ci_lower_95", "ci_upper_95", "p_raw"),
        skip_absent = TRUE
    )

    # FDR (per model ID)
    if ("p_raw" %in% names(all_results)) {
        all_results[, p_fdr := p.adjust(p_raw, method = "BH"), by = model_id]
        all_results[, fdr_significant := (p_fdr < 0.05)]
    }

    # Metadata
    all_results[, `:=`(
        model_family = "FE-continuous",
        evidence_tier = "confirmatory",
        practical_flag = (abs(effect_size) >= 0.03) # Standardized threshold assumption
    )]

    out_path <- file.path(args$out_dir, "heterogeneity_coeffs_r.csv")
    fwrite(all_results, out_path)
    cat("Saved R results to", out_path, "\n")
} else {
    cat("No results generated.\n")
}
